src/runCompletion.ts

```diff
@@ -35,7 +50,57 @@ export default async function runCompletion(
 
   const result = await autocomplete(requestData, timeout);
 
-  return result;
+  if (result?.results.length || !retry?.cancellationToken) {
+    return result;
+  }
+
+  return handleRetries(requestData, retry);
+}
+
+function handleRetries(
+  requestData: AutocompleteParams,
+  {
+    cancellationToken,
+    interval = 200,
+    timeout = 1000,
+  }: {
+    cancellationToken?: CancellationToken;
+    interval?: number;
+    timeout?: number;
+  }
+): Promise<AutocompleteResult | null | undefined> | null | undefined {
+  if (cancellationToken?.isCancellationRequested) {
+    return null;
+  }
+  return new Promise((resolve, reject) => {
+    let timeoutId: NodeJS.Timeout | undefined;
+    const intervalId = setInterval(() => {
+      void autocomplete({ ...requestData, cached_only: true })
+        .then((result) => {
+          if (result?.results.length) {
+            clearInterval(intervalId);
+            clearTimeout(timeoutId as NodeJS.Timeout);
+            resolve(result);
+          }
+        })
+        .catch((error) => {
+          clearInterval(intervalId);
+          clearTimeout(timeoutId as NodeJS.Timeout);
+          reject(error);
+        });
+    }, interval);
+
+    timeoutId = setTimeout(() => {
+      clearInterval(intervalId);
+      resolve(null);
```

Comment:
before in this case you did not get null. You would get `{ results: [] }`

Maybe keep that same behavior?

Reply:
how about this https://github.com/codota/tabnine-vscode/pull/1092/files#diff-1ea23267f59e70ebf2cdb6ff62d4a004b59126d651954b36436758bf1c603cd2R97

Reply:
commit https://github.com/codota/tabnine-vscode/pull/1092/commits/cd3f5d84a6e93cc04599a7132dc00b258d97beaa